### ISSUE TITLE
Report diagnostic for use of implicit culture-sensitive ToString()?

### ISSUE THREAD

**Issue by drieseng** [meziantou/Meziantou.Analyzer#894]
You currently have **MA0076** to report a diagnostic when an implicit culture-sensitive `ToString(...)` is used in an interpolated string.

As far as I'm aware, there's no diagnostic for reporting use of a culture-sensitive `ToString(...)` in other contexts.
I reckon it may be very costly to run such a rule against a large code base, so it may not be the best idea to implement this (unless not enabled by default).

```c#
using System;

internal sealed class Program
{
    private static void Main(string[] args)
    {
        Foo foo = new Foo();

        Console.WriteLine(foo); // No diagnostic is reported
        Console.WriteLine($"{foo}"); // MA0076 is reported
    }
}

public sealed class Foo : IFormattable
{
    public string ToString(string format)
    {
        return string.Empty;
    }

    public string ToString(string format, IFormatProvider formatProvider)
    {
        return string.Empty;
    }

    public override string ToString()
    {
        return string.Empty;
    }
}
```

**Comment by meziantou** [comment]
- MA0011: Report when using `ToString()` and an overload with a `IFormatProvider` is available
- MA0075 / MA0076: Report in string interpolation and concatenation

I think what MA0076 is currently missing is implicit conversion to string. 

````c#
var foo = new foo();
string v = foo; // using an implicit conversion
````

However, that would not cover `Console.WriteLine(foo)` because of the numerous overloads (e.g. `Console.WriteLine(object)`). However, I'm not sure how to better detect this case as `Console.WriteLine` is not the only method/type having this kind of overloads. I'm not sure how many methods would need to be special cased. Detecting `WriteLine` method name would not be valid, for instance `TextWriter.WriteLine` is ok as the format is set as a property.

Also, `Console.WriteLine` is not a very useful method in code that goes in production. So, is it worth handling it?

**Comment by drieseng** [comment]
My earlier example was bogus.
This is a better example:

```c#
using System;

internal sealed class Program
{
    private static void Main(string[] args)
    {
        Foo foo = new Foo();

        string s = foo; // No diagnostic produced

        Console.WriteLine(foo);  // No diagnostic produced
        Console.WriteLine(s);
    }
}

public sealed class Foo : IFormattable
{
    public string ToString(string format)
    {
        return string.Empty;
    }

    public string ToString(IFormatProvider formatProvider)
    {
        return string.Empty;
    }

    public string ToString(string format, IFormatProvider formatProvider)
    {
        return string.Empty;
    }

    public override string ToString()
    {
        return string.Empty;
    }

    public static implicit operator string(Foo d)
    {
        return string.Empty;
    }
}
```

You wouldn't need to special case any methods.
You only need to report the diagnostic if the type also implements **IFormattable** or defines a `ToString(IFormatProvider)` method.

**Comment by drieseng** [comment]
That said, I'm sure you have more important things to work on.
I'm ok with closing or postponing this issue.

**Comment by meziantou** [comment]
I'm not sure of this one. If the type has a conversion operator, then I assume an explicit design choice has been made by the developer. So, I don't know if there is a need for reporting it.

**Comment by drieseng** [comment]
Ok, let's forget about this one. Sorry if I wasted your time.
